### PR TITLE
Huge ffmpeg hack: compile x86_gcc2 FFmpeg with GCC 7

### PR DESCRIPTION
**Pull request by waddlesplash**
Tested with GCC2 MediaPlayer; it does work.

See individual commits, because the overall diff probably won't be too helpful. Revision bumps done using the new [revision bumper script](https://github.com/haikuports/haikuporter/commit/92e916366d7fbb48d2e04a169874ca21436dcb11).

### CODE REVIEW

File: media-video/ffmpeg/ffmpeg-4.0.1.recipe [hunk 1]
```diff
@@ -55,25 +56,29 @@ PROVIDES="
...
 	lib:libvorbis$secondaryArchSuffix
 	lib:libvorbisenc$secondaryArchSuffix
-	lib:libvpx$secondaryArchSuffix
 	lib:libz$secondaryArchSuffix
 	"
+if [ "$effectiveTargetArchitecture" != x86_gcc2 ]; then
+	REQUIRES+="
+		lib:libvpx$secondaryArchSuffix
+		lib:libgnutls$secondaryArchSuffix
+	"
```
Comment (fbrosson) [inline media-video/ffmpeg/ffmpeg-4.0.1.recipe]: I guess you meant:
```diff
 	REQUIRES+="
 		lib:libvpx$secondaryArchSuffix
 		lib:libgnutls$secondaryArchSuffix
-	"
+		"
```
Hmm, instead, you could, maybe, remove one tab in lines 69, 70 and 71, and keep line 72 unchanged.

(Same remarks for the closing double quotes (at line 165) of the `BUILD_REQUIRES` bloc.)

BTW, I haven't tested your hack yet, but it seems it is an excellent idea.

Regarding the way this PR will be merged, I think "**Rebase** and merge" is the right thing to do, and I guess you will be using this method, (instead of "Squash and merge").

I would, however amend the very first commit of this PR:
* first, revert the `sed s/ffmpeg/ffmpeg4/` @extrowerk did in `PROVIDES{,_avdevice,_devel,_tools}`;
* second, remove `ffmpeg-4.0.recipe` and `ffmpeg-4.0.patchset`.

This will allow to have slightly smaller `git diff` for both the first and second commits of this PR.
And the "**Rebase** and merge" method will be perfect for this PR.

**Comment by waddlesplash** [comment]
Haiku tickets this fixes:
 * https://dev.haiku-os.org/ticket/13981
 * https://dev.haiku-os.org/ticket/12436
 * https://dev.haiku-os.org/ticket/5080
 * https://dev.haiku-os.org/ticket/13410
 * https://dev.haiku-os.org/ticket/13337
 * https://dev.haiku-os.org/ticket/8461

Fixed, but perhaps not by this:
 * https://dev.haiku-os.org/ticket/8605
 * https://dev.haiku-os.org/ticket/8511

Probably fixed but couldn't test:
 * https://dev.haiku-os.org/ticket/13989

**Comment by waddlesplash** [comment]
Haiku tickets this fixes:
 * https://dev.haiku-os.org/ticket/13981
 * https://dev.haiku-os.org/ticket/12436
 * https://dev.haiku-os.org/ticket/5080
 * https://dev.haiku-os.org/ticket/13410
 * https://dev.haiku-os.org/ticket/13337
 * https://dev.haiku-os.org/ticket/8461
 * https://dev.haiku-os.org/ticket/12696

Partially fixed:
 * https://dev.haiku-os.org/ticket/11272

Fixed, but perhaps not by this:
 * https://dev.haiku-os.org/ticket/8605
 * https://dev.haiku-os.org/ticket/8511
 * https://dev.haiku-os.org/ticket/6984

Probably fixed but couldn't test:
 * https://dev.haiku-os.org/ticket/13989

**Comment by korli** [comment]
Not sure this fixes tickets which also already affected x86_64, how would you explain this outcome?

**Comment by waddlesplash** [comment]
Dunno. At least some of the ones that were said to occur on x64 (e.g. mod playing is broken) were indeed reproducible with this patch. I just retested all the "broken codecs" ones and those are what was fixed.

File: media-video/ffmpeg/ffmpeg-4.0.1.recipe [hunk 2]
```diff
@@ -123,10 +128,27 @@ CONFLICTS_devel="
 	ffmpeg2${secondaryArchSuffix}_devel
 	"
 
+# -- HACK! --
+# ffmpeg ~0.10 has perpetually had & caused problems when compiled with GCC 2.
+# (Later versions do not compile with GCC 2 at all.) So instead, we now compile
+# FFmpeg for x86_gcc2 with a modern GCC.
+#
+# This works because GCC's C ABI has not changed between 2 and 7; only the C++
```
Comment (pulkomandy) [inline media-video/ffmpeg/ffmpeg-4.0.1.recipe]: There are no guarantees, and indeed you had to hack some libgcc support.

File: media-video/ffmpeg/ffmpeg-4.0.1.recipe [hunk 3]
```diff
@@ -136,12 +158,19 @@ BUILD_REQUIRES="
 	devel:libvpx$secondaryArchSuffix
```
Comment (pulkomandy) [inline media-video/ffmpeg/ffmpeg-4.0.1.recipe]: Did you forget to remove this?

File: media-video/ffmpeg/ffmpeg-4.0.1.recipe [hunk 4]
```diff
@@ -156,9 +185,14 @@ PATCH()
 	sed -i "s,/usr/bin/perl,$portPackageLinksDir/cmd~perl/bin/perl," \
 		doc/Doxyfile \
 		doc/texi2pod.pl
+	# force no-undefined-symbols
+	if [ "$effectiveTargetArchitecture" = x86_gcc2 ]; then
+		sed -i "s/SHFLAGS='-shared/SHFLAGS='-shared -Wl,--no-undefined/g" \
```
Comment (pulkomandy) [inline media-video/ffmpeg/ffmpeg-4.0.1.recipe]: Why is this needed? And wy only on gcc2?

File: media-video/ffmpeg/ffmpeg-4.0.1.recipe (hunk 3 above)
Comment (waddlesplash) [reply media-video/ffmpeg/ffmpeg-4.0.1.recipe]: Ah, I did. Oops.

File: media-video/ffmpeg/ffmpeg-4.0.1.recipe (hunk 2 above)
Comment (waddlesplash) [reply media-video/ffmpeg/ffmpeg-4.0.1.recipe]: Well, there are no "guarantees", but the C ABI hasn't changed in over 20 years, it probably won't now. :)

File: media-video/ffmpeg/ffmpeg-4.0.1.recipe (hunk 4 above)
Comment (waddlesplash) [reply media-video/ffmpeg/ffmpeg-4.0.1.recipe]: Because this will then error at build time instead of runtime if GCC uses some libgcc function we don't have.

**Comment by waddlesplash** [comment]
https://git.io/fSGqG
